Try: parallelize CI jobs

sirreal opened the pull request:
CI has the concept of "workflows", which allow "jobs", or sequential sets of discrete "steps", to be run in parallel.

This change reorganizes our CircleCI to use a workflow with the following setup:

- Setup: Handle repo checkout and npm installation.
  - Translate
  - Test

Translate and test "require" setup and will be run in parallel, as different jobs, when setup is complete. You can see [workflow run examples‘(https://circleci.com/workflow-run/8703ecb5-ac98-401b-a49a-6e475e112ec1) on this branch:

![workflow](https://user-images.githubusercontent.com/841763/45430940-d4740600-b6a6-11e8-9c35-f434ab863519.png)

This change opens the door for repos with independent test suites to be folded into the Calypso monorepo while maintaining an independent test suite that can be run in parallel. In particular, the `i18n` integration: #27123 

You can see that, although some jobs are run in parallel, the workflow run is not significantly faster than a normal sequential run. However, if we expect independent test suites to be introduced, the parallelization will provide speed benefits as well as isolation of which job (test suite) has failures.

## Other

We were caching `.git` so that checkouts are sped up, however we'd never update them because a CI cache is never overwritten. This adds branch and revision to the cache key so that it _will_ be updated and a recent cache will be used.

We were using the `-browsers` node CI image, which includes browsers. That does not appear to be used in any way, so I've changed to the standard CI node image.

## Testing
- This should not _slow down_ CI runs
- The test suites run should be the same
- Translations work as expected

sirreal committed:
Try parallelized CI

sirreal committed:
Update and restore CircleCI git cache

sirreal committed:
Add headers

sirreal committed:
Try running tests

sirreal committed:
Npx

sirreal committed:
Hack the runner

sirreal committed:
Remove i18n-related work

sirreal committed:
Parallelize translation and testing

sirreal committed:
Store the translation artifacts

sirreal committed:
Drop browsers image

sirreal committed:
Better name for workflow

blowery commented on .circleci/config.yml:
```diff
@@ -1,56 +1,91 @@
+  # Build translations
+  #
+  translate:
+    parallelism: 6
```
i think this can just be `1`. translate isn't parallelized.

sirreal replied:
Yes!

sirreal replied:
0d978540fc84c1ad99b49d515e32df97b8ec913c

sirreal committed:
Translate isn't parallel

sirreal committed:
Share artifact and result setup

sirreal committed:
DROPME: Demo test failure

sirreal committed:
Try better test reporting

sirreal committed:
Clean up Jest CI command

sirreal commented on .circleci/config.yml:
```diff
@@ -1,56 +1,95 @@
           environment:
-              NODE_ENV: test
+            NODE_ENV: test
           command: npm run build-server
```
Is this a requirement for translations? Why do we build the server?

sirreal replied:
#7665 @seear 

It looks like this was an SSR thing, but with canaries we might just drop it?

I thought it might be related to i18n, but after looking at history I'm doubting that.

seear replied:
It was to ensure that the SSR pragma check was run over the codebase during Circle CI tests. That checking was removed in #9585, so we will likely no longer need to build the server here.

sirreal replied:
Great! Ripped out in 039b15c

sirreal committed:
Only maxWorkers

sirreal committed:
Tweak jest more

sirreal committed:
Json not js

sirreal committed:
Remove build-server from CircleCI tests

sirreal commented on .circleci/config.yml:
```diff
@@ -1,57 +1,90 @@
+      - run: *setup-results-and-artifacts
 
       - run:
           name: Build calypso-strings.pot
```
If this isn't parallel, we can drop the `if [ "$CIRCLE_NODE_INDEX" == "0" ]` bits 🎉

sirreal replied:
✅ 66bcfbb

DanReyLop commented on .circleci/config.yml:
```diff
@@ -69,6 +102,23 @@ jobs:
+  # Test Calypso
+  #
+  test-calypso:
+    parallelism: 6
```
According to CircleCI, our plan has 10 containers. Should we decrease `parallelism` to `5` so 2 PRs can run at the same time, or is CircleCI smart enough about that?

sirreal replied:
I'm not sure about our plan or how it works, but there are jobs that are happy to use free containers.

There are plenty of significant improvements we can make in this PR and tweak this number later.

sirreal replied:
We might even play with this then use insights to look at build and queue time over an extended period.

sirreal committed:
Remove redundant node index checks

This job is no longer parallel so the checks shouldn't run

sirreal committed:
Use jest-junit over jest-junit-reporter

sirreal commented:
With the removal of `build-server`, translate is running in less time than persist/attach workspace.

There are some other improvements here, but until there are more test suites to run, parallelizing these steps is likely a new slowdown.

blowery commented:
still one failing test... seems odd. https://circleci.com/gh/Automattic/wp-calypso/104043#tests/containers/5

sirreal committed:
Add eslint enable/disable comments

sirreal committed:
Try runInBand

sirreal committed:
Revert "Try runInBand"

This reverts commit 2e0e46073c2f886e3c1e6b58b83f547b75b11fcf.

sirreal committed:
Move server test invocation into CircleCI config

sirreal committed:
Remove noop integration tests step from CI

sirreal committed:
Remove redundant gather tests step

Test results are now written directly to the expected location and do
not need to be moved.

dmsnell commented on .circleci/config.yml:
```diff
@@ -1,79 +1,121 @@
-             - v1-npmcache-{{ checksum ".nvmrc" }}
-
+      - restore_cache: *restore-npm-cache
       - run: npm ci
```
does restoring the cache here simply mean that some packages won't have to download? I thought that performing `npm ci` would mean we can ignore caches because it would wipe out any existing packages

sirreal replied:
In #27180 this changed to cache `node_modules` directly, skipping the npm installation on a cache hit. Cache matches checksums on both `.nvmrc` (should correspond to Node.js version in the container) and `npm-shrinkwrap.json`. From #27180:

> Add `node_modules` cache. Previously we had eliminated the `node_modules` cache because we were using `npm ci` which deletes the folder (#25431). Then, we maintained this behavior, but began to cache the npm cache so that `npm ci` would execute faster because the modules were already available locally (#25487). This approach restores the `node_modules` cache and _completely skips_ `npm ci` if there was a cache hit. This should be safe because the `node_modules` cache requires a checksum match on `.nvmrc` and `npm-shrinkwrap.json`. It's important to ensure that _all variables_ which might generate different `node_modules` structure are part of the cache key, e.g. running on different architectures.

dmsnell commented on .circleci/config.yml:
```diff
@@ -1,79 +1,121 @@
-              npm run translate; mkdir -p $CIRCLE_ARTIFACTS/translate
-              mv calypso-strings.pot $CIRCLE_ARTIFACTS/translate
-            fi
+            npm run translate; mkdir -p $CIRCLE_ARTIFACTS/translate
```
is there a reason these are on the same line but the next command has its own?

sirreal replied:
No. In #27180 the directory creation is removed here and added to the shared directory creation.

dmsnell commented on .circleci/config.yml:
```diff
@@ -1,79 +1,121 @@
-              bash gp-localci-client/generate-new-strings-pot.sh $CIRCLE_BRANCH $CIRCLE_SHA1 $CIRCLE_ARTIFACTS/translate
-              rm -rf gp-localci-client
-            fi
+            git clone https://github.com/Automattic/gp-localci-client.git
```
would it be helpful here to perform a shallow clone? I don't think we need the history do we?

```bash
git clone --depth=1 https://github.com/Automattic/gp-localci-client.git
```

Would we have any benefit to caching this and doing a fetch instead?

sirreal replied:
> would it be helpful here to perform a shallow clone? I don't think we need the history do we?

😁 I agree (#26226):

https://github.com/Automattic/wp-calypso/blob/f851182623c77178af75b51dda621aeed6ec84b8/.circleci/config.yml#L65

> Would we have any benefit to caching this and doing a fetch instead?

The repo we're checking out is small enough that the entire step often completes with a duration of `0:00`. It's not really worth optimizing:

![zero](https://user-images.githubusercontent.com/841763/45803217-4c4ebb80-bcb9-11e8-8d55-0f9c4c0e5e43.png)

I wish the repo didn't even exist. It's just a bash script that's shared between a couple of projects.

dmsnell commented on .circleci/config.yml:
```diff
@@ -1,79 +1,121 @@
+      - run: *setup-results-and-artifacts
 
       - run:
           name: Lint Config Keys
```
something like these lint jobs could also be done in parallel. is there any benefit in assigning them to separate jobs vs. using the inherent `parallelism` here?

sirreal replied:
There's some overhead for almost all of our jobs because they all require some setup and teardown (repo checkout, npm install, store test results). There's a base cost for running pretty much any of our jobs.

It's also important to economize containers. We have a limited container pool, and if we use them, other jobs are queued. If setup and teardown takes ~10s and we imagine some short jobs where the principal steps take 2s, both jobs might finish earlier if they're parallelized, but considering overhead it's a net loss:

Parallelized jobs: 10s overhead + 2s steps = 12s * 2 = 24s
Single job: 10s overhead + 2s steps + 2s steps = 14s

If the jobs start at the same moment, the parallel workflow might finish 2s earlier at 12s, but it uses 24s of container time. In the single job, we trade 2s of clock time but free up 10s of container tim.

The lint steps are generally very fast because we're only those that differ from master (#25494). I think there's good reason to keep them together. In #27180, the lint and translation steps are all part of one job.

The other reason to split jobs is that they can be selectively run via the workflow with filters for branches, tags, cron schedules, manual approval. For example, a tag or manual approval workflow might be an interesting idea to explore in #27282

sirreal replied:
In the context of this PR, there's good reason to isolate the client tests from any other steps because the client tests are by far the most time consuming. #27180 does just that.

dmsnell commented on .circleci/config.yml:
```diff
@@ -87,29 +129,48 @@ jobs:
+                --reporters=default                   \
+                --reporters=jest-junit                \
+                --config=test/client/jest.config.json \
+                --maxWorkers=2                        \
```
why `parallelism: 6` and `maxWorkers=2`?

sirreal replied:
`parallelism` are the containers that will run this job in parallel. It is defined for the job, so 6 containers will run each step of this job. 6 seemed to work well, but it's something that we should tweak in the future. If job queues tend to form, we might try lowering this a bit. I'd expect that with lower parallelism jobs would take a bit longer, but we'd have more containers free to run more jobs at the same time.

---

`maxWorkers` defines the maximum size of the worker pool Jest uses to run tests in parallel. CircleCI [recommands](https://circleci.com/docs/2.0/collect-test-data/#jest) `--runInBand`, which is analogous to `--maxWorkers=1`, but I believe those docs are outdated.  I did not seem improvements with `runInBand` or issues when using 2 workers.

Me theory for why 2 workers per container is a sweet spot is the [`resource_class`](https://circleci.com/docs/2.0/configuration-reference/#resource_class). We're using the default, so we'll have 2 vCPUs, one for each worker. You can confirm the container resources at the top of a job:

![resources](https://user-images.githubusercontent.com/841763/45802924-72c02700-bcb8-11e8-8636-5e61baabd6d3.png)

dmsnell reviewed (approved):
@sirreal what do we need to make this happen?

this is great work, I appreciate the clean naming and logical flow to this setup.

please let me know if there's something you need help with. even though I left comments I feel this is good enough to merge.

sirreal commented:
Sorry @dmsnell, this was initial exploratory work that was polished and prepared in #27180. I should have closed this so that was clear 😞 

Some of your comments are still valid in the context of #27180, so I'll reply.
